### PR TITLE
Don't cache the Compilation object in steps in the interop source generators

### PR DESCRIPTION
**Pull request by jkoritzinsky**
The interop source generators today use a custom Equals implementation to ignore the Compilation object that is stored in the StubEnvironment object as part of the "incremental" data model. Our steps are still incremental, but this change can cause an indeterminate amount of Compilation objects to be kept alive significantly longer, which increases memory load in VS and the Roslyn OOP service. This change makes one scenario slightly less incremental (it will regenerate the source for some stubs when the file a stub is in is changed), but this will overall be an improvement in memory usage for our generators.

Extracted out of #68276

It might be worth pushing this PR into 7.0 or 7.0.1 due to the VS impact.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGeneratorHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.Interop
+{
+    internal static class ComInterfaceGeneratorHelpers
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGeneratorHelpers.cs:12]: Is this file intentionally added?

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGeneratorHelpers.cs:12]: The name is confusing given our complete lack of support fro COM interfaces :)

File: src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/IncrementalGenerationTests.cs [hunk 2]
```diff
@@ -210,5 +209,52 @@ public async Task ChangingMarshallingAttributes_SameStrategy_DoesNotRegenerate()
...
+            Compilation comp1 = await TestUtils.CreateCompilation(new[] { syntaxTree });
+
+            var (reference, driver) = RunTwoGeneratorOnTwoIterativeCompilationsAndReturnFirst(comp1);
+
+            GC.Collect();
+
+            Assert.False(reference.IsAlive);
+            GC.KeepAlive(driver);
+
+            static (WeakReference reference, GeneratorDriver driver) RunTwoGeneratorOnTwoIterativeCompilationsAndReturnFirst(Compilation startingCompilation)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/IncrementalGenerationTests.cs:229]: Just for sanity, perhaps we should mark this as "no inline". Given the size this is unlikely, but an implied invariant here is that this not be inlined.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SequenceEqualImmutableArray.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Microsoft.Interop
+{
+    public readonly record struct SequenceEqualImmutableArray<T>(ImmutableArray<T> Array, IEqualityComparer<T> Comparer)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SequenceEqualImmutableArray.cs:11]: Please document this type and why it is needed. My naive understand here is the override of `Equals` but I'm missing the dots between why this helps but the removal of the other custom `Equals` is no longer needed.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGeneratorHelpers.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGeneratorHelpers.cs:12]: No, this snuck in as part of the cherry-pick. I'll remove it.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SequenceEqualImmutableArray.cs (hunk 3 above)
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SequenceEqualImmutableArray.cs:11]: Will do

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by pavelsavara — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
Since the backport failed, I'll handle backporting this manually.

To make sure we cover all cases of the "keeping the Compilation object alive" problem, I suggest we wait to backport this until https://github.com/dotnet/runtime/issues/78242 is also fixed.

**Pull request merged**